File: leetcode/101-200/173-binary-search-tree-iterator/bench/bst_iterator.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    long val;
    long left;
    long right;
} Node;

static Node *nodes;
static long node_count;

static long insert(long root, long val) {
    if (root == -1) {
        nodes[node_count].val = val;
        nodes[node_count].left = -1;
        nodes[node_count].right = -1;
        return node_count++;
    }
    if (val < nodes[root].val) {
        long li = nodes[root].left;
        nodes[root].left = insert(li, val);
    } else {
        long ri = nodes[root].right;
        nodes[root].right = insert(ri, val);
    }
    return root;
}

int main(void) {
    long n = 4000, passes = 30000;
    nodes = malloc(n * sizeof(Node));
    node_count = 0;
    long root = -1;
    long state = 12345;
    for (long i = 0; i < n; i++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long v = state % 1000;
        root = insert(root, v);
    }

    long *stack = malloc(n * sizeof(long));
    long sink = 0;
    for (long p = 0; p < passes; p++) {
        long idx = (p * 1315423911L + 7L) % n;
        nodes[idx].val = (nodes[idx].val + 1) % 1000;

        long sp = 0;
        long cur = root;
        while (cur != -1) {
            stack[sp++] = cur;
            cur = nodes[cur].left;
        }
        long pos = 1;
        while (sp > 0) {
            long top = stack[--sp];
            sink += pos * nodes[top].val;
            pos++;
            long r = nodes[top].right;
            while (r != -1) {
                stack[sp++] = r;
                r = nodes[r].left;
            }
        }
    }
    printf("%ld\n", sink);
    free(stack); free(nodes);
    return 0;
}

File: leetcode/101-200/173-binary-search-tree-iterator/bench/bst_iterator.rs
struct Node {
    val: i64,
    left: i64,
    right: i64,
}

fn insert(nodes: &mut Vec<Node>, root: i64, val: i64) -> i64 {
    if root == -1 {
        nodes.push(Node { val, left: -1, right: -1 });
        return nodes.len() as i64 - 1;
    }
    if val < nodes[root as usize].val {
        let li = nodes[root as usize].left;
        let l = insert(nodes, li, val);
        nodes[root as usize].left = l;
    } else {
        let ri = nodes[root as usize].right;
        let r = insert(nodes, ri, val);
        nodes[root as usize].right = r;
    }
    root
}

fn main() {
    let n: i64 = 4000;
    let passes: i64 = 30000;

    let mut nodes: Vec<Node> = Vec::new();
    let mut root: i64 = -1;
    let mut state: i64 = 12345;
    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        let v = state % 1000;
        root = insert(&mut nodes, root, v);
    }

    let mut stack: Vec<i64> = Vec::new();
    let mut sink: i64 = 0;
    for p in 0..passes {
        let idx = ((p * 1315423911 + 7) % n) as usize;
        nodes[idx].val = (nodes[idx].val + 1) % 1000;

        stack.clear();
        let mut cur = root;
        while cur != -1 {
            stack.push(cur);
            cur = nodes[cur as usize].left;
        }
        let mut pos: i64 = 1;
        while !stack.is_empty() {
            let top = stack.pop().unwrap();
            sink += pos * nodes[top as usize].val;
            pos += 1;
            let mut r = nodes[top as usize].right;
            while r != -1 {
                stack.push(r);
                r = nodes[r as usize].left;
            }
        }
    }
    println!("{}", sink);
}
